what4/src/What4/Solver/CVC5.hs

```diff
@@ -0,0 +1,241 @@
+import           What4.Solver.Adapter
+import           What4.Utils.Process
+
+
+intWithRangeOpt :: ConfigOption BaseIntegerType -> Integer -> Integer -> ConfigDesc
+intWithRangeOpt nm lo hi = mkOpt nm sty Nothing Nothing
+  where sty = integerWithRangeOptSty (Inclusive lo) (Inclusive hi)
+
+data CVC5 = CVC5 deriving Show
+
+-- | Path to cvc5
+cvc5Path :: ConfigOption (BaseStringType Unicode)
+cvc5Path = configOption knownRepr "solver.cvc5.path"
+
+cvc5RandomSeed :: ConfigOption BaseIntegerType
+cvc5RandomSeed = configOption knownRepr "solver.cvc5.random-seed"
+
+-- | Per-check timeout, in milliseconds (zero is none)
+cvc5Timeout :: ConfigOption BaseIntegerType
+cvc5Timeout = configOption knownRepr "solver.cvc5.timeout"
+
+-- | Control strict parsing for cvc5 solver responses (defaults
+-- to solver.strict-parsing option setting).
+cvc5StrictParsing :: ConfigOption BaseBoolType
+cvc5StrictParsing = configOption knownRepr "solver.cvc5.strict_parsing"
+
+cvc5Options :: [ConfigDesc]
+cvc5Options =
+  let pathOpt co = mkOpt co
+                   executablePathOptSty
+                   (Just "Path to CVC5 executable")
+                   (Just (ConcreteString "cvc5"))
+      p1 = pathOpt cvc5Path
+      r1 = intWithRangeOpt cvc5RandomSeed (negate (2^(30::Int)-1)) (2^(30::Int)-1)
+      tmOpt co = mkOpt co
+                 integerOptSty
+                 (Just "Per-check timeout in milliseconds (zero is none)")
+                 (Just (ConcreteInteger 0))
+      t1 = tmOpt cvc5Timeout
+  in [ p1, r1, t1
+     , copyOpt (const $ configOptionText cvc5StrictParsing) strictSMTParseOpt
+     ] <> SMT2.smtlib2Options
+
+cvc5Adapter :: SolverAdapter st
+cvc5Adapter =
+  SolverAdapter
+  { solver_adapter_name = "cvc5"
+  , solver_adapter_config_options = cvc5Options
+  , solver_adapter_check_sat = runCVC5InOverride
+  , solver_adapter_write_smt2 = writeCVC5SMT2File
+  }
+
+indexType :: [SMT2.Sort] -> SMT2.Sort
+indexType [i] = i
+indexType il = SMT2.smtlib2StructSort @CVC5 il
+
+indexCtor :: [SMT2.Term] -> SMT2.Term
+indexCtor [i] = i
+indexCtor il = SMT2.smtlib2StructCtor @CVC5 il
+
+instance SMT2.SMTLib2Tweaks CVC5 where
+  smtlib2tweaks = CVC5
+
+  smtlib2arrayType il r = SMT2.arraySort (indexType il) r
+
+  smtlib2arrayConstant = Just $ \idx rtp v ->
+    SMT2.arrayConst (indexType idx) rtp v
+  smtlib2arraySelect a i = SMT2.arraySelect a (indexCtor i)
+  smtlib2arrayUpdate a i = SMT2.arrayStore a (indexCtor i)
+
+  smtlib2declareStructCmd _ = Nothing
+  smtlib2StructSort []  = Syntax.varSort "Tuple"
+  smtlib2StructSort tps = Syntax.Sort $ "(Tuple" <> foldMap f tps <> ")"
+    where f x = " " <> Syntax.unSort x
+
+  smtlib2StructCtor args = Syntax.term_app "mkTuple" args
+  smtlib2StructProj _n i x = Syntax.term_app (Syntax.builder_list ["_", "tupSel", fromString (show i)]) [ x ]
+
+cvc5Features :: ProblemFeatures
+cvc5Features = useComputableReals
+           .|. useLinearArithmetic
+           .|. useNonlinearArithmetic
```

Comment:
These two entries are redundant, as they are [implied by `useIntegerArithmetic` and `useComputableReals`](https://github.com/GaloisInc/what4/blob/e569514d10f940db20027705388565f3d9f729e3/what4/src/What4/ProblemFeatures.hs#L70-L76).